src/Storage/StorageObject.php

```diff
@@ -633,6 +636,178 @@ public function downloadAsStream(array $options = [])
     }
 
     /**
+     * Create a Signed URL for this object.
+     *
+     * Please note that using this method requires that the OpenSSL extension be
+     * available in your PHP installation. In most cases, it will be by default,
+     * but failures can often be rectified by installing OpenSSL.
+     *
+     * Example:
+     * ```
+     * $url = $object->signedUrl(new Timestamp(new DateTime('tomorrow')));
+     * ```
+     *
+     * ```
+     * // Create a signed URL allowing updates to the object.
+     * $url = $object->signedUrl(new Timestamp(new DateTime('tomorrow')), [
+     *     'method' => 'PUT'
+     * ]);
+     * ```
+     *
+     * @see http://php.net/manual/en/book.openssl.php OpenSSL
+     *
+     * @param Timestamp $expires A Timestamp specifying when the URL will expire.
+     * @param array $options {
+     *     Configuration Options.
+     *
+     *     @type string $method One of `GET`, `PUT` or `DELETE`. **Defaults to**
+     *           `GET`.
+     *     @type string $cname The CNAME for the bucket, for instance
+     *           `https://cdn.example.com`. **Defaults to**
+     *           `https://storage.googleapis.com`.
+     *     @type string $contentMd5 The MD5 digest value in base64. If you
+     *           provide this, the client must provide this HTTP header with
+     *           this same value in its request. If provided, take care to
+     *           always provide this value as a base64 encoded string.
+     *     @type string $contentType If you provide this value, the client must
+     *           provide this HTTP header set to the same value.
+     *     @type array $headers If these headers are used, the server will check
+     *           to make sure that the client provides matching values. Provide
+     *           headers as a key/value array, where the key is the header name,
+     *           and the value is the header value.
+     *     @type string $saveAsName The filename to prompt the user to save the
+     *           file as when the signed url is accessed. This is ignored if
+     *           `$options.responseDisposition` is set.
+     *     @type string $responseDisposition The
+     *           [`response-content-disposition`](http://www.iana.org/assignments/cont-disp/cont-disp.xhtml)
+     *           parameter of the signed url.
+     *     @type string $responseType The `response-content-type` parameter of the
+     *           signed url.
+     *     @type array $keyFile Keyfile data to use in place of the keyfile with
+     *           which the client was constructed. If `$options.keyFilePath` is
+     *           set, this option is ignored.
+     *     @type string $keyFilePath A path to a valid Keyfile to use in place
+     *           of the keyfile with which the client was constructed.
+     * }
+     * @return string
+     */
+    public function signedUrl(Timestamp $expires, array $options = [])
```

Comment:
> What is the benefit to accepting a Timestamp vs \DateTimeImmutable? Or, why not allow both? WDYT about accepting a few other input options, such as a unix timestamp and string format?

Done!

> Should we provide a factory timestamp method like we do for other libraries that utilize the Timestamp type?

I believe so.

> Thoughts on breaking this up into a few private methods to improve readability?

I'm not sure whether it would improve things much. Right now it's kind of long, but it's top-to-bottom. There's nothing that's used elsewhere so moving to private methods may just end up requiring more cursor moving to read the code.